src/utils/SpaceFillingTree.cpp

```diff
@@ -0,0 +1,266 @@
+/** Copyright (C) 2017 Tim Kuipers - Released under terms of the AGPLv3 License */
+#include "SpaceFillingTree.h"
+
+namespace cura {
+
+
+SpaceFillingTree::SpaceFillingTree(
+    Point middle,
+    coord_t radius,
+    unsigned int depth)
+: root(nullptr)
+, depth(depth)
+{
+    unsigned int root_depth = 0;
+    root = new Node(nullptr
+    , root_depth
+    , middle
+    , Direction::DIRECTION_COUNT // The root has no parent to child direction, so DIRECTION_COUNT is used as exception value
+    , depth
+    );
+
+    // total width = radius becase 1 + .5 + .25 + ... = 2
+    // therefore the initial width = .5 * radius
+    coord_t first_offset = radius / 2; // horizontal/vertical distance from the middle to the end of the first line segment
+    // construct 1st order children
+    root->construct(first_offset);
+    root->prune();
+    root->distance_depth = 0;
+    root->setDistanceDepth();
+}
+
+
+SpaceFillingTree::SpaceFillingTree(SpaceFillingTree&& other)
+: aabb(other.aabb)
+, root(other.root)
+{ // move constructor
+    other.root = nullptr;
+}
+
+SpaceFillingTree::~SpaceFillingTree()
+{
+    if (root)
+    {
+        delete root;
+    }
+}
+
+void SpaceFillingTree::walk(SpaceFillingTree::LocationVisitor& visitor) const
+{
+    root->walk(visitor);
+}
+
+void SpaceFillingTree::debugOutput(SVG& out, bool output_dfs_order) const
+{
+    out.writePolygon(aabb.toPolygon());
+    int root_order = 0;
+    root->debugOutput(out, root->middle, output_dfs_order, root_order); // root will draw line from its middle to the same middle
+}
+
+void SpaceFillingTree::debugCheck() const
+{
+#ifdef DEBUG
+    root->debugCheck();
+#endif // DEBUG
+}
+
+SpaceFillingTree::Node::Node(SpaceFillingTree::Node* parent, unsigned int recursion_depth, Point middle, Direction parent_to_here_direction, unsigned int total_depth)
+: parent(parent)
+, recursion_depth(recursion_depth)
+, total_depth(total_depth)
+, middle(middle)
+, parent_to_here_direction(parent_to_here_direction)
+{
+    for (Node*& child_place : children)
+    {
+        child_place = nullptr;
```

Comment:
No it would only be zero-initialized when in the constructor it would say `: children({})`.
I find that kind of notation not really clear, though.

That's case 3 in your linked page.

https://stackoverflow.com/questions/38892455/initializing-an-array-of-zeroes